linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala

```diff
@@ -43,121 +43,113 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
-          for (rhs <- desugarExpr(jsSuperClass, resultType = AnyType)) yield {
-            js.VarDef(fileLevelVar("superClass").ident, Some(rhs))
-          }
+    if (!tree.kind.isJSClass) {
+      assert(tree.jsSuperClass.isEmpty, className)
+
+      if (useClasses) {
+        val parentVarWithGlobals = for (parentIdent <- tree.superClass) yield {
+          implicit val pos = parentIdent.pos
+          if (shouldExtendJSError(tree)) globalRef("Error")
+          else WithGlobals(globalVar("c", parentIdent.name))
         }
 
-        for {
-          optStoreJSSuperClass <- WithGlobals.option(genStoreJSSuperClass)
-          entireClassDef <- entireClassDefWithGlobals
-          createStaticFields <- genCreateStaticFieldsOfJSClass(tree)
-        } yield {
-          tree.jsClassCaptures.fold {
-            val classValueIdent = fileLevelVarIdent("b", genName(className))
-
-            val createClassValueVar = genEmptyMutableLet(classValueIdent)
-
-            val createAccessor = {
-              val classValueVar = js.VarRef(classValueIdent)
-
-              val body = js.Block(
-                  js.If(!classValueVar, {
-                    js.Block(
-                        optStoreJSSuperClass.toList :::
-                        entireClassDef ::
-                        createStaticFields :::
-                        (classValueVar := globalVar("c", className)) ::
-                        genClassInitialization(tree)
-                    )
-                  }, {
-                    js.Skip()
-                  }),
-                  js.Return(classValueVar)
-              )
+        for (parentVar <- WithGlobals.option(parentVarWithGlobals)) yield {
+          globalClassDef("c", className, parentVar, allES6Defs)
+        }
+      } else {
+        allES5Defs(globalVar("c", className))
+      }
+    } else {
+      // Wrap the entire class def in an accessor function
+      import TreeDSL._
 
-              globalFunctionDef("a", className, Nil, body)
-            }
+      val genStoreJSSuperClass = tree.jsSuperClass.map { jsSuperClass =>
+        for (rhs <- desugarExpr(jsSuperClass, resultType = AnyType)) yield {
+          js.VarDef(fileLevelVar("superClass").ident, Some(rhs))
+        }
+      }
 
-            js.Block(createClassValueVar, createAccessor)
-          } { jsClassCaptures =>
-            val captureParamDefs = for (param <- jsClassCaptures) yield {
-              implicit val pos = param.pos
-              val ident = fileLevelVarIdent("cc", genName(param.name.name),
-                  param.originalName.orElse(param.name.name))
-              js.ParamDef(ident, rest = false)
-            }
+      val classValueIdent = fileLevelVarIdent("b", genName(className))
+      val classValueVar = js.VarRef(classValueIdent)
+
+      val entireClassDefWithGlobals = if (useClasses) {
+        genJSSuperCtor(tree).map { jsSuperClass =>
+          classValueVar := js.ClassDef(None, Some(jsSuperClass), allES6Defs)
```

Comment:
We should still provide the `classValueIdent` as the `name` of the `js.ClassDef`, even when used in expression position. This makes sure that its `.name` property will be correctly filled in by the engine, and that overall the class and its methods appear in a better way in debugging tools.